added new Plugin for Wemos Motorshield P_078

TungstenE2 opened the pull request:
added new Plugin for Wemos Motorshield P_078

Tested already on Wemos D1 mini

update: P_078 was changed to P_079, see below.

TungstenE2 committed:
Added Wemos Motorshield Plugin (No. 78)

TungstenE2 committed:
Added Wemos Motorshield Plugin (No. 78)

TungstenE2 commented:
Do you need any further details or input for this pull request?
Plugin is working without problems so far.

TD-er commented:
I need a little time to make it "compatible" with the new way of keeping track of the plugin-specific settings.
Now you can only use 1 of these plugins on the same board.
This also means the command for the motor board should be extended a little bit, which is being prepared by @workgroupengineering (see PR #1522 ).

I could merge it now, but then we have to document the current commands, users start using it, then we change the command, projects stop working when upgraded, etc. :)

So that's why it takes some time to merge.

TungstenE2 commented:
thx for the update

TungstenE2 commented:
any news? would be cool to have it in the main project.

TD-er commented:
I will look at it this weekend.


Edit 2018/07/23 01:13 am.
I knew I forgot something... didn't get to it to work on this one too :(

TungstenE2 commented:
push... ;-)

TD-er commented:
I have 3 remarks about the PR:

1. The layout of the code is a bit messy (lines starting at start of line, completely ignoring indents)
2. The use of `delay(100)` I am trying to get all plugins to be non blocking and 100 msec waiting will block the node for other jobs. (for example 5 runs of 50/sec plugin calls)
3. Why not use the (new) I2C classes to interface with the chip? Now I have to check all calls to see if they do start and end the communications like they should.

TD-er committed:
[MotorShield] Rename P078 to P079

TD-er committed:
[MotorShield] Cleanup code layout

Indentation was hard to read, so ran code beautify.

TD-er committed:
[MotorShield] Added comments in the code and fixes

Frequency was not using the upper 4 bits of the values.
Changed delay to yield.

TD-er commented:
I did some cleaning and removed some bugs.

TungstenE2 commented:
thx a lot! I was busy the last weeks and my friend who helped me was on sick leave.

TungstenE2 commented:
what are the next steps?

will the plugin be added to the plugin list? who will do this?
https://www.letscontrolit.com/wiki/index.php/Official_plugin_list

do I need to write a wiki like this one for the Adafruit motorshield?
https://www.letscontrolit.com/wiki/index.php/%22Adafruit_Motor_Shield_v2%22

Grovkillen commented:
I can do the update of the list. I don't have the hardware myself so I'm not best suited to write the wiki for it. If you apply for wiki rights, make sure to check your spam folder for the credentials. Or send me the text if you don't want to do it.

TungstenE2 commented:
no worries, I can do it. just need some days... ;-)

TungstenE2 commented:
@Grovkillen account was created, but I have no rights to create new wiki page. who can grant me create privileges?

Grovkillen commented:
I can create the wiki page and let you do the actual text.

TungstenE2 commented:
@Grovkillen ok, thx. please send me the link.

TungstenE2 commented:
https://www.letscontrolit.com/wiki/index.php/WemosMotorshield

initial draft

TungstenE2 commented:
just tested the plugin on 'Wemos D1 mini' and 'Wemos D1 mini Pro' with ESP_Easy_mega-20180826_test_ESP8266_4096.bin

works perfect with my DC motor at the chicken door! ;-)

TungstenE2 commented:
I am just curious when plugins will be moved from testing to normal status? Is there a process or timeline?

sparkescreative commented:
Hi Guy's

Longtime user of ESP Easy but I'm struggling to get this plugin to work. I am also using it for a chicken coop door in this situation.

I have v2 of the module
![motor_v2 0 0_2_16x16](https://user-images.githubusercontent.com/22837930/93105861-77e9be00-f6a7-11ea-98ac-abfba58fafba.jpg)
![motor_v2 0 0_1_16x16](https://user-images.githubusercontent.com/22837930/93105873-7c15db80-f6a7-11ea-9d68-d22f9546800a.jpg)

I'm using the latest version of ESP Easy and a NodeMCU

I have not updated the firmware on the module to solve the I2C problem others have mentioned as mine was listed after the first i2c scan and hasn't changed or dropped of at all. So not been an issue.

I can send commands that the log seems to list and accept. I get and OK in the command line output.

But my motor does nothing. I have an external power source connected to the motor shield that can power the motor directly without a problem.

Can anybody shed any light.

I have noticed that when this modules is connected to a Wemos D1 they connect the RST on the shield module to the RST of the mini.

Should I be doing something with the RST pin?

TD-er commented:
The RST pad looks like it is meant for connecting RST to GPIO-16.
That's needed if you use deep sleep, to wake the ESP.
So if you don't use deep sleep, you don't need to do anything with that pad.

Can you check using a multimeter if the outputs of the A1 has some output voltage compared to GND?

Also please check to see if this unit has the correct firmware loaded.
If my memory serves me wel, there are some motor shields out there (not 100% sure it is this one) which were shipped with a buggy firmware.
It was also causing issues when just scanning for I2C devices.  It would then lock up the I2C bus.

TungstenE2 commented:
I only have v1 shield, and wiki documatation was based on this. May be check with TomasB. He updated the wiki for v2.

TD-er commented:
Then you must refer to him, or else I guess he won't feel summoned :)
@thomastech

sparkescreative commented:
"Can you check using a multimeter if the outputs of the A1 has some output voltage compared to GND?"

No theres no voltage there.

On another note any ideas why there is a 3.3v & 5V inputs to the module. Do I need both connected. I've been running it at 3.3v?

sparkescreative commented:
> Also please check to see if this unit has the correct firmware loaded.
> If my memory serves me wel, there are some motor shields out there (not 100% sure it is this one) which were shipped with a buggy firmware.
> It was also causing issues when just scanning for I2C devices. It would then lock up the I2C bus.

From the Wiki it says the firmware was intended for v1 and I have v2.

TD-er commented:
I don't know the shield, but I do see a "VM" pad next to GND, which seems to be a separate voltage input for the motor driver. (which totally makes sense to have)

So I would expect you must connect the power supply for the motor to those VM and GND pin.

But I'm not entirely sure, so please try to find a schematic of the board.

TD-er commented:
If this board only supports 2 motors, I think you should measure a voltage on the pins A1, A2 (positive  when turning in one direction, negative in the other direction)

sparkescreative commented:
Right I've found the schematic
[https://docs.wemos.cc/en/latest/_static/files/sch_motor_v2.0.0.pdf](url)

And your right about VM and GND along the bottom and it's how I have it connected. Separate power for the motor.

sparkescreative commented:
This is how I have it wired

![Screen Shot 2020-09-14 at 19 38 25](https://user-images.githubusercontent.com/22837930/93124717-f18ea580-f6c1-11ea-8392-b44597f815b3.png)

thomastech commented:
You'll need to connect the RST pin too. This ensures reliable power-up initialization of the motor shield. Other than that the wiring diagram seems legit.
**EDIT: Heads-up, I just noticed that D1 and D2 wiring is swapped in the image.** 

Still doesn't work? 
(1) Even if you already done it a thousand times, double check your wiring again. Could be miss-wired, solder shorts, etc.
(2) Post a screenshot of the Motorshield's device page that shows your settings. Make sure it is "Enabled."
(3) Show us the complete syntax of the commands you are sending to test it out.

- Thomas

sparkescreative commented:
Hi Thomas

> You'll need to connect the RST pin too. This ensures reliable power-up initialization of the motor shield. Other than that the wiring diagram seems legit.

I don't want to sound stupid but do I connect the RST on the shield with the RST on the NodeMCU?

Thank you by the way


![Screen Shot 2020-09-14 at 20 53 00](https://user-images.githubusercontent.com/22837930/93131750-7d0d3400-f6cc-11ea-86c8-59e2d898515e.png)
![Screen Shot 2020-09-14 at 20 53 33](https://user-images.githubusercontent.com/22837930/93131758-7e3e6100-f6cc-11ea-9012-2468e4826031.png)

thomastech commented:
Wiring: NodeMCU RST to Wemos RST.

The "Ok" reply in the Command Output is meaningless, so don't use it to confirm the command was successfully processed. No error checking is performed, it only confirms that the "WemosMotorShieldCmd" name was found in the command library. In other words, you could disconnect the Wemos Motor board, remove the Wemos Plugin, and still get a Ok reply.

- Thomas

sparkescreative commented:
I don't want to sound stupid but do I connect the RST on the shield with the RST on the NodeMCU?

sparkescreative commented:
There is a little bit of current between the GND and the motor Pins but only 0.020 of a volt.

This doesn't change when I run a command like "wemosMotorShieldCMD,0,backward,255"

Is there anything in the Debug log that might help?

thomastech commented:
Q.  ... do I connect the RST on the shield with the RST on the NodeMCU?
A.  Wemos Motor Shield RST to NodeMCU RST. 

I noticed that your motor command is sending the speed value 255. This value represents PWM % and should range from 0 to 100. But I recall that another user could not get 100 to work and used a much lower value instead to turn on his motor. So maybe try:
`WemosMotorShieldCmd,0,Forward,30`

- Thomas

sparkescreative commented:
Sorry I missed your answer the first time about the RST. I've connect them now. Restarted and I've just tried the command `WemosMotorShieldCmd,0,Forward,30` 

Unfortunatly that didn't work... It obviously hates me or I've got a crap board. I'm going to check all the solder connections and wires again....

Thank you again for you help with this.

thomastech commented:
Sorry to hear it still doesn't work. Could be a bad board, bug in the latest plugin, or evil spirits.

Just for grins, send the motor command using http from a browser. The URL would be something like this:
`http://<ESP IP address>/control?cmd=WemosMotorShieldCMD,0,Forward,20`

If still no luck, then try an older ESPEasy Mega release. Searching the forum I found a user that had success using the mega-20200426 release with his Wemos Motor shield. So try that version.

- Thomas

sparkescreative commented:
That my friend is a wicked answer.

I tried the url instead it gave me the awesome "ok" but still didn't work.

I've just given up for this evening and gone off to shoot cowboys in Red Dead Redemption. But tomorrow I will do what you suggested and try and older firmware. I'll let you know...... fingers crossed

sparkescreative commented:
I've flash mega-20200426 on to a brand new NodeMCU this morning and unfortunately I still can't get the motor shield to do anything. I must have a dodgy board.

sparkescreative commented:
Well I've tested the shield with a default test sketch from Wemos and the shield worked fine and the motor moved.

So it must be something to do with ESPEasy or something I'm doing wrong with the setup!?!?!?!?

sparkescreative commented:
This is the sketch I used

/* Simple Motor Shield Test. */

#include <Wire.h>
#include <LOLIN_I2C_MOTOR.h>

LOLIN_I2C_MOTOR motor; //I2C address 0x30
// LOLIN_I2C_MOTOR motor(DEFAULT_I2C_MOTOR_ADDRESS); //I2C address 0x30
// LOLIN_I2C_MOTOR motor(your_address); //using customize I2C address

void setup()
{
  Serial.begin(115200);
  Serial.println("Motor Shield Testing...");

  while (motor.PRODUCT_ID != PRODUCT_ID_I2C_MOTOR) //wait motor shield ready.
  {
    motor.getInfo();
  }
}

void loop()
{
  Serial.println("Change A to CCW, B to CW, Freq: 1000Hz");
  Serial.println("Duty Tesing...");

  motor.changeFreq(MOTOR_CH_BOTH, 1000); //Change A & B 's Frequency to 1000Hz.
  /*
      motor.changeFreq(MOTOR_CH_A, 1000);//Change A 's Frequency to 1000Hz.
      motor.changeFreq(MOTOR_CH_B, 2000);//Change B 's Frequency to 2000Hz.
  */
  motor.changeStatus(MOTOR_CH_A, MOTOR_STATUS_CCW);
  motor.changeStatus(MOTOR_CH_B, MOTOR_STATUS_CW);

  for (float duty = 0; duty <= 100; duty++)
  {
    motor.changeDuty(MOTOR_CH_A, duty);
    motor.changeDuty(MOTOR_CH_B, 100 - duty);

    Serial.print("Change A Duty to ");
    Serial.print(duty);
    Serial.println("%");

    Serial.print("Change B Duty to ");
    Serial.print(100 - duty);
    Serial.println("%");

    delay(100);
  }

  Serial.println("STANDBY Tesing...");
  motor.changeStatus(MOTOR_CH_BOTH, MOTOR_STATUS_CCW);
  motor.changeDuty(MOTOR_CH_BOTH, 100);
  for (int i = 0; i < 5; i++)
  {
    Serial.println("MOTOR_STATUS_STANDBY");
    motor.changeStatus(MOTOR_CH_BOTH, MOTOR_STATUS_STANDBY);
    delay(500);
    Serial.println("MOTOR_STATUS_CW");
    motor.changeStatus(MOTOR_CH_BOTH, MOTOR_STATUS_CW);
    delay(500);
    Serial.println("MOTOR_STATUS_STANDBY");
    motor.changeStatus(MOTOR_CH_BOTH, MOTOR_STATUS_STANDBY);
    delay(500);
    Serial.println("MOTOR_STATUS_CCW");
    motor.changeStatus(MOTOR_CH_BOTH, MOTOR_STATUS_CCW);
    delay(500);
  }

  Serial.println("MOTOR_STATUS Tesing...");
  for (int i = 0; i < 5; i++)
  {
    Serial.println("MOTOR_STATUS_STOP");
    motor.changeStatus(MOTOR_CH_A, MOTOR_STATUS_STOP);
    delay(500);
    Serial.println("MOTOR_STATUS_CCW");
    motor.changeStatus(MOTOR_CH_A, MOTOR_STATUS_CCW);
    delay(500);
    Serial.println("MOTOR_SHORT_BRAKE");
    motor.changeStatus(MOTOR_CH_A, MOTOR_STATUS_SHORT_BRAKE);
    delay(500);
    Serial.println("MOTOR_STATUS_CW");
    motor.changeStatus(MOTOR_CH_A, MOTOR_STATUS_CW);
    delay(500);
  }
}

sparkescreative commented:
> Sorry to hear it still doesn't work. Could be a bad board, bug in the latest plugin, or evil spirits.
> 
> Just for grins, send the motor command using http from a browser. The URL would be something like this:
> `http://<ESP IP address>/control?cmd=WemosMotorShieldCMD,0,Forward,20`
> 
> If still no luck, then try an older ESPEasy Mega release. Searching the forum I found a user that had success using the mega-20200426 release with his Wemos Motor shield. So try that version.
> 
> * Thomas

Just found this Thomas

Is the final comment yours?

I think I have the same problem

https://www.letscontrolit.com/forum/viewtopic.php?f=6&t=2844&p=40481#p14464

TD-er commented:
Just to be sure, can you check what the I2C scanner shows with this board connected?

And can you also please check with a lower I2C clock rate? (e.g. 100 kHz)

thomastech commented:
@TD-er: It looks like we're beating a dead horse here. If you follow the link in the previous post, the Wemos Motor plugin doesn't support the V2 board. There's no mentions of this in the ESPEasy docs. Oh the humanity!
BTW, if you review the V1 and V2 libraries you will see that they are substantially different. 

@sparkescreative: Unless TD-er has a solution (try his suggestions), the best I can offer is to update the ESPEasy Motorshield plugin wiki and have it state that V2 boards are NOT supported.

- Thomas

TungstenE2 commented:
@thomastech the plugin was invented for v1 shield. Any reference added to the wiki pointing to v2 was done by user ThomasB. This should be rolled back then.

sparkescreative commented:
@TD-er I tried 100kHz and 400kHz. It shows on both.

https://user-images.githubusercontent.com/22837930/93241412-36c3dd80-f77d-11ea-9747-65b4b64c3472.png


@thomastech  I had a look at the libraries today when I realised the board was using a different one. They are very different. 

I did try to update it myself to the v2 library but I'm not that skilled... lol

It looks like this plugin may well be obsolete if you can't buy the version 1 module. I tried today!

Would it be worth updating to the v2 library? I understand that is a big ask.

TD-er commented:
> Would it be worth updating to the v2 library? I understand that is a big ask.

If they are very different, I guess it would make more sense to make a new plugin for the v2.
No need to include the v1 version in a build if the v2 is needed and vice verse.

But maybe we could still share some code like the command syntax, etc.

Open for suggestions here :)

thomastech commented:
@TungstenE2: I am ThomasB. The only previous edit I did was to add corrected PCB Solder Jumper images of the V1 board, plus a short comment that the V2 board does not require the jumpers. 

I will revise the V2 comment to specifically say that the V2 board is not supported. I will also fix the Wiki's stated maximum motor speed range for the V1 board, which is currently listed as 255 (0-255). But according to the source code it is hard limited to 100.

- Thomas

sparkescreative commented:
@TungstenE2  The WeMos v2 should work because it uses the same library but it seems to have been replaced by a Lolin v2 which has the updated library.

So basically

Wemos = Good
Lolin = Bad

@thomastech A new plugin would be the way to go. So we would end up having seperate Wemos and Lolin versions.

thomastech commented:
@TungstenE2: The Lolin V2 is just a newer version of the original Wemos V1 board. I believe that Wemos and Lolin are branding names from the same business entity. From what I've seen, newer board releases seem to be getting the Lolin name instead of Wemos. Sometimes both names are printed on the same board.

@TD-er: For the least amount of end-user confusion, I think expanding the existing plugin to support both board versions would be ideal. That is to say, include a check box to select the board type.

- Thomas

TD-er commented:
The Wemos brand was at some point bought by Lolin.
But on the other hand, I've seen way too many boards that also claim to have Wemos and/or Lolin and in various spelling variations which do not look like the photo's on the Lolin site.
So it does look like there are some forgeries sold online.
Not saying this board _is_ a forgery, but I'm always a bit skeptical when I see "Wemos" branded boards being sold.

I guess this may come as a shock to some, but forgeries do appear when a brand name gets popular.
I'm still waiting to see the first "ESPEasy" branded boards to be sold via Ali Express/Banggood/Ebay, so maybe we need to spend some more time on branding :)

sparkescreative commented:
I'm quite sure mine is an original Lolin. I studied there photo's and every think matches. I also bought it from a reputable seller. 

I'm happy to offer my design and branding skills. 

My company is https://www.matthewsparkes.co.uk

TD-er commented:
Well all jokes aside, I think we should have a look at how much code differences there are and then decide on either creating a new plugin, or do the sensible thing and follow @thomastech 's advice on UX :)

thomastech commented:
Given that the V2 board being debugged here is working with the official V2 library, I'd say it is legit hardware. The problem is that the V2 is not compatible with ESPEasy.

To use the Lolin V2 board a generous developer will need to revise the existing Motor plugin or create a dedicated "Lolin V2" plugin. So flip a coin to choose the solution then prepare to bribe a developer with caffeinated beverages or craft brews.

@TD-er: BTW, have you upgraded to Platformio to 5.0.1 yet? I did and the ESPEasy projects are causing weird things to happen whenever I run the library update task from any project. I haven't had time to dig deeper, but the problems go away when I remove my ESPEasy projects. So that's what I've done for now.

- Thomas

TD-er commented:
I have updated to PIO 5.0.1 and indeed the upgrade wasn't as smooth as it could be.

One of the things I noticed was that it would wait forever to update when it could not find the Python3 binaries, even though they were installed and added to the path.

Also there is a mayor change in PIO5 which I really don't like.
If you 'unfold' an environment, PIO will generate the file ESPEasy.ino.cpp and delete it when done.
This is regardless of an active build which does fail when the .cpp is deleted.

Also it no longer can find our own core 2.4.2 with PUYA patches. So maybe other libraries also experience strange issues.

sparkescreative commented:
@thomastech Happy to insert coffee - just direct me to the right cup!

You guy's are all wicked in my book

thomastech commented:
@TD-er: Sounds like PIO 5.0.1 is adding interesting (hair pulling) challenges to our lives. At the moment it is working fine for me, but I had to remove all my ESPEasy projects to make it play nice.

@sparkescreative: Regarding a coffee contribution, I'd suggest directing any bribes at TD-er.  He seems to be the best man for the job. Or if his daughter's coding skills are up to it, then maybe some premium apple juice for her sippy cup?

- Thomas

sparkescreative commented:
> @sparkescreative: Regarding a coffee contribution, I'd suggest directing any bribes at TD-er. He seems to be the best man for the job. Or if his daughter's coding skills are up to it, then maybe some premium apple juice for her sippy cup?

@TD-er Would this be something you would have time to do? I can defiantly send coffee/homebrew

TD-er commented:
Well Lilly's coding skills are not yet at the level we require for this specific plugin.
She only knows how to write the alfabeth and numbers and such, but last week we had some debate about negative numbers and she became somewhat angry as she thought I made fun of her when giving the answer to her question of 14 - 20.

So if you would bribe her, then the chances are your motor will only turn in 1 direction.

sparkescreative commented:
@TD-er My little lady loves taking stuff apart. I gave her an old Kodak Phaser printer and it kept her busy for a whole weekend. Obviously I couldn't find any of my screwdrivers afterwards but I now have got a bunch of motors and sensors I can use for not projects.

Looks like I have to bribe you then sir... What do I have to do?

TD-er commented:
Well I also have to order the board, as I am not going to debug via "can you try this and let me know so I can add a new debug statement in the next build"

So can you at least give the link to the exact board (preferrably from the original Lolin store)

sparkescreative commented:
@TD-er Of course.

Board
https://docs.wemos.cc/en/latest/d1_mini_shiled/motor.html

Schematic
https://docs.wemos.cc/en/latest/_static/files/sch_motor_v2.0.0.pdf

Library
https://github.com/wemos/LOLIN_I2C_MOTOR_Library

Examples (I tested my board with one of these)
https://github.com/wemos/LOLIN_I2C_MOTOR_Library/tree/master/examples

The Shop link on their site takes you here
https://de.aliexpress.com/item/32700182142.html?spm=a2g0o.store_home.productList_2473300.pic_2

thomastech commented:
@TD-er: It's certainly a good idea for you to have both MotorShield boards for your ongoing ESPEasy software validation. But perhaps I can help reduce your workload. While the boards are in transit from the China Supplier I'll attempt to add the MotorShield V2 compatibility. 

I have the V1 and V2 MotorShield boards and should be able to find some time this week to work on the plugin. If you're already deep into coding it then no worries, it's all yours.  Just let me know if I should proceed.

Fine print: Keep in mind that if I fail you would revert to the original plan. So don't put off ordering the boards.

- Thomas

TD-er commented:
If you can start doing the coding and debugging, that would be great.

thomastech commented:
Ok, I'm on it.
- Thomas

thomastech commented:
@sparkescreative: I'm nearly done with the V2 validation and should have a bin for you to try in a couple hours. What build type do you need? 

If uncertain, just report the name of the existing bin and I can determine the build I need to create for your device.

Also, it would be a very good idea if one or more existing Wemos V1 users tested the new pluging to make sure their existing configuration still works correctly. Looking for volunteers ...

- Thomas

thomastech commented:
Status Update: I've finished the validation of a [normal] build. Both boards are working on my test bed that uses a Wemos D1 Mini 4MB.

@TD-er: The uncrustify.cfg file I have is one you created about a year ago. I couldn't find it in latest distribution (mega-20200913). Can you point me to where I can download your latest?

- Thomas

TD-er commented:
@thomastech 
It should be there, in the source tree: https://github.com/letscontrolit/ESPEasy/blob/mega/uncrustify.cfg

sparkescreative commented:
@thomastech Epic Thomas. I'm not sure what you mean by "bin" but I use the latest mega release and I compile it myself with Arduino on a mac. Flashing it onto a NodeMCU v3.

Don't I just need the plugin file? Sorry if I'm being a dumb.

thomastech commented:
@TD-er: Thanks. My ESPEasy project came from the latest zip file downloaded from the Releases page. Uncrustify is missing from it. For example: https://github.com/letscontrolit/ESPEasy/releases/tag/mega-20200913
Fine print: Yes, I should have built from the github distro. But the zip method is less hassle for me.

@sparkescreative: If you can build your bin from source code then that is great. Here is the beta release file:
<File deleted, see newer version below>

See the file's header for the change log. Some notable things:
- In addition to the WemosMotorShieldCMD command keyword, you can also use LolinMotorShieldCMD.
- Command syntax errors are reported in the browser and info log.
- You can test communication to your V2 board. Send LolinMotorShieldCMD without parameters.

Thomas

sparkescreative commented:
@thomastech Perfect. I'm having a bash now and I'll let you know what happens.

sparkescreative commented:
@thomastech Just one error for me.

/ESP8266 Projects/ESPEasy_ESP82xx_mega-20200812/source/ESPEasy/_P079_Wemos_Motorshield.ino: In function 'boolean Plugin_079(byte, EventStruct*, String&)':
_P079_Wemos_Motorshield:210:10: error: 'PLUGIN_WEBFORM_SHOW_I2C_PARAMS' was not declared in this scope
     case PLUGIN_WEBFORM_SHOW_I2C_PARAMS:
          ^

'PLUGIN_WEBFORM_SHOW_I2C_PARAMS' was not declared in this scope

Have I got to change something somewhere else as it's a new file?

sparkescreative commented:
@thomastech Ignore previous comment. I've sorted the firmware - I wasn't using the latest release - so I'm going to test the motor shield now...... Panic over

thomastech commented:
PLUGIN_WEBFORM_SHOW_I2C_PARAMS is a new Plugin state function. I think it appeared at the end of Aug 2020. So as you found, using the latest source is the fix to the compile error.

- Thomas

sparkescreative commented:
@thomastech Worked first time. I've now got a bunch of motors spinning around in my studio. You've earned beer/coffee/gold without a doubt.

The only negative is that the wiki will need updating.

thomastech commented:
That's good news. Thanks for quickly testing it out.

I'm putting the finishing touches on the plugin (code cleanup). I'll post the final version for you to try out. Should be ready in a couple hours.

I'll update the wiki too. I don't have access to editing readthedocs, so someone else gets to update that doc.

- Thomas

thomastech commented:
@sparkescreative Here is the revised plugin. Please try it out to confirm nothing was messed up.
[_P079_Wemos_Motorshield.zip](https://github.com/letscontrolit/ESPEasy/files/5248577/_P079_Wemos_Motorshield.zip)
Edit: Uploaded newer file.

- Thomas

sparkescreative commented:
@thomastech It's all good. I've tried all the commands on both motor outputs and everything works as expected. Looks like its good to be updated in the master branch. Well done. You got skills.

I've just implemented it this morning in my chicken coop. Now I don't have to get up at sunrise to let my chickens out. It controls a DC motor that opens the door just after sunrise and closes it at sunset. It already logs the coops Temperature and Humidity so now it logs the position of the door and when its next going to open or close.

This is how it looks logged to my platform:
![Screen Shot 2020-09-19 at 12 34 35](https://user-images.githubusercontent.com/22837930/93666256-93bcdd80-fa74-11ea-8baa-2976d18af938.png)

On a side note - Is there a plugin for stepper motor support?

TungstenE2 commented:
very nice, this is my Tablet UI of the chicken house. 

![grafik](https://user-images.githubusercontent.com/39955685/93667222-6aa44900-fa84-11ea-8cbd-c6f56555837d.png)

sparkescreative commented:
> very nice, this is my Tablet UI of the chicken house.
> 

@TungstenE2 Wicked. Is that your own interface? Are you logging it historically?

A camera was next on my list after an automatic food/water refill system.

I have a few ESP32-CAM's flashed with some real basic custom firmware but they are crap when it's dark. I haven't found a way to get ESP Easy onto them and utilise the camera yet. But I want too.

Here are some more screenshots of my platform:
![iPhone-Screenshots](https://user-images.githubusercontent.com/22837930/93667982-86f2b680-fa81-11ea-910d-b2efcdd0535a.jpg)

TD-er commented:
@thomastech 
ReadTheDocs is also included in the mega branch.
See the docs/ dir.
The used syntax is Sphinx.

Are you going to make a PR, or else give me a notice when a zip file can be downloaded.

sparkescreative commented:
@TD-er Hi Sir. Quick question. Is there a plugin for stepper motor support?

TD-er commented:
The [Adafruit motor shield](https://www.adafruit.com/products/1438) does seem to have stepper support.  (P048)
So 2 H-bridges of the board are then combined to drive a single stepper motor.

thomastech commented:
@TD-er: I will make a PR for this after one last review with my fresh eyes. 

But before creating the PR I was hoping that a user would install it on an existing Wemos V1 installation to confirm that it is a seamless transition. 
@TungstenE2 You're the best candidate for this. Especially since you created the Wemos plugin and know how it should behave. Plus you're using it with a custom web interface.

Because I'm the one that touched this plugin I will update the wiki and readthedocs. I'll do that after the plugin is released.

- Thomas

TungstenE2 commented:
@sparkescreative no, I am using FHEM as controller. This is the Tablet UI for FHEM.

@thomastech as my v1 board is in the chicken house I can only update via OTA. I can do a test this afternoon, hoping this does not break anything. Please provide a .bin file to be uploaded via OTA.
Currently I am using: ESP_Easy_mega_20200721_normal_ESP8266_4M1M.bin

thomastech commented:
@TungstenE2 thank you for being a volunteer! Here is the bin: 
[ESP_Easy_mega_20200920_normal_ESP8266_4M1M.zip](https://github.com/letscontrolit/ESPEasy/files/5252103/ESP_Easy_mega_20200920_normal_ESP8266_4M1M.zip)

BTW, I added support for the "brake" command on this test release. That could be useful on some ESPEasy applications.

- Thomas

TungstenE2 commented:
Moin, I can confirm that updated plugin is working with v1 board.
I was able to open and close via android app. Rules are working.

![grafik](https://user-images.githubusercontent.com/39955685/93737161-ba008b80-fbe2-11ea-805f-9702f8497093.png)

![grafik](https://user-images.githubusercontent.com/39955685/93737244-f338fb80-fbe2-11ea-8fd6-6be1f70e1364.png)

![grafik](https://user-images.githubusercontent.com/39955685/93737264-fcc26380-fbe2-11ea-8578-2dce14af7072.png)

thomastech commented:
Thank you for performing the test, I appreciate it. It's comforting to know the plugin is working in your installation.

I'm currently editing the readthedocs files for inclusion in the PR. I'm using the reStructuredText extension in Platformio to do this. Installing the extension was an interesting ride (several hours of hair pulling and fist fighting).

- Thomas

TD-er commented:
There is a requirements.txt file in the docs dir.
So `pip install -r requirements.txt` should do the trick.

thomastech commented:
@TD-er Thanks for the tip. It reports "Requirement already satisfied" for each item in the requirements.txt pip installation, so it looks like I was already setup.  Fortunately, after the installation fight it has been working and readthedocs editing is moving along.

-> Besides the rst files, does the PR need to include the regenerated HTML output too? It seems doubtful you would need it, but thought I should ask.

- Thomas

TD-er commented:
Nope, only the source files.
The build output is also in the .gitignore file, so you should not be able to append it to a commit.

thomastech commented:
@TD-er: The PR for P079 has been submitted. Fingers are crossed that I did it correctly.

The PR includes updated readthedocs. I'll edit the wiki after the new plugin goes live. 

- Thomas

TD-er commented:
Initial scan of the PR does look good.
Tomorrow I will have a second look and if it still looks good, it will be merged soon :)

thomastech commented:
Thanks for reviewing it. If you need me to make any changes then just let me know.

- Thomas

thomastech commented:
@TD-er: It's great to see the new Motorshield plugin in today's 20200929 release.

Looking at the change log, I saw that now there's support for multiple instances of the I2C related plugins. Is using this documented somewhere? 

I'm mostly interested in how it handles plugins (like this one) that respond to a fixed command keyword. I suspect you added new command syntax that prevents all instances from responding to the same keyword. So it would be interesting to review the docs and see how it's done.

- Thomas

tonhuisman commented:
@thomastech in PR #3210 I briefly introduced the features/configuration of the I2C multiplexer support. Included in that PR is the link to the new Hardware page docs, including the multiplexer support. 😄

thomastech commented:
@tonhuisman: Thanks for pointing to the docs.

It's still not clear to me how to communicate to a particular multiplexed I2C device. Lets say I have to send commands like these:

```
WemosMotorShieldCMD,<Motornumber>,<Action>,<Speed>
or
oledframedcmd,<line>,<text>

```
If there are multiple I2C instances of those two devices, how would I direct the command to a particular device?

TD-er commented:
The commands still have to be dealt with, that's another one of @tonhuisman 's open PR's (he's a busy man)
See [multiplexer documentation](https://espeasy.readthedocs.io/en/latest/Hardware/Hardware.html?highlight=multiplexer#i2c-multiplexer) for more ... well ... documentation :)

And if you scroll down. you'll also see documentation on Ethernet support.

Just as a teaser, this is a board I made for a client, but there will also be an open source board design for ESP32 + LAN which can be produced by yourself (and JLCPCB for the really small parts, which are so small even their pick-and-place machine had some issues with it)

![image](https://user-images.githubusercontent.com/3751318/94602454-2b9ca180-0295-11eb-81b3-d625664f1501.png)

TD-er commented:
> @tonhuisman: Thanks for pointing to the docs.
> 
> It's still not clear to me how to communicate to a particular multiplexed I2C device. Lets say I have to send commands like these:
> 
> ```
> WemosMotorShieldCMD,<Motornumber>,<Action>,<Speed>
> or
> oledframedcmd,<line>,<text>
> ```
> 
> If there are multiple I2C instances of those two devices, how would I direct the command to a particular device?

See: https://github.com/letscontrolit/ESPEasy/pull/3198

thomastech commented:
Ahh, got it. I appreciate that the prefix global name will be optional. That would allow existing installations to be updated without requiring modifications to rules or plugins.

- Thomas
